[WIP] VINN Update

LeHenschel opened the pull request:
This branch updates the network to the newest VINN version (Henschel L, Kügler D, Reuter M. FastSurferVINN: Building resolution-independence into deep learning segmentation methods-A solution for HighRes brain MRI. Neuroimage. 2022 May 1;251:118933. doi: 10.1016/j.neuroimage.2022.118933. Epub 2022 Feb 3. PMID: 35122967.). Running the old CNN model is still possible as well. The commits include significant updates to the code base, including loading/saving of config files, new classes for integration with tensorboard, logging and optimizer/lr-scheduler selection. 

Recently tested for running predictions on HCP-TRT as well as training on the original multi-resolution dataset from the publication.

Still missing are updates to the README (run commands and usage explanation), checkpoints, corresponding Docker files, containers, requirements.txt as well as Tutorials. 

Also still required is integration into recon-surf with the updated hires pipeline and final testing (TRT and Group Analysis)

LeHenschel committed:
configuration defaults for VINN

LeHenschel committed:
Gaussian Noise augmentation and ZeroPadding added

LeHenschel committed:
Added functionalitz to read classes from lut and changed mapping to torch version

LeHenschel committed:
Datasets for loading multi-res hdf5-files

LeHenschel committed:
Hdf5-generation for multi-resolution images

LeHenschel committed:
Dataloader set up script

LeHenschel committed:
added new VINN architecture and corresponding building blocks

LeHenschel committed:
Function to automatically select optimizer and loss-functions

LeHenschel committed:
Function for loading and saving checkpoints

LeHenschel committed:
Load, save and overwrite config files

LeHenschel committed:
Automatic log-setup

LeHenschel committed:
Variety of learning rate scheduler to select from

LeHenschel committed:
Tensorboard logging procedure

LeHenschel committed:
Dice class for logging to tensorboard

LeHenschel committed:
Plotting functions

LeHenschel committed:
Model evaluation class

LeHenschel committed:
Model training class

LeHenschel committed:
Run script for model training

LeHenschel committed:
Run script for model prediction

LeHenschel committed:
Added license info

LeHenschel committed:
Merge branch 'feature/vinn' of https://github.com/Deep-MI/FastSurfer into feature/vinn
Pull branch for improvements

LeHenschel committed:
Fixed import issues for modules

LeHenschel committed:
Fixed default classes to saggital number of classes (51)

LeHenschel committed:
Fixed prediction run, removed unncessary arguments

LeHenschel committed:
Changed num workers to 8

LeHenschel committed:
Changed kernel H and W to correct default of 3

LeHenschel committed:
Fix check for empty argument

LeHenschel committed:
Remove obsolete smooth param

LeHenschel committed:
Fixed import issue

LeHenschel committed:
Fixed naming of vinn

LeHenschel committed:
Deleted files

LeHenschel committed:
Remove wrong hdf5-gen script

LeHenschel committed:
Updated hdf5-gen script for multi-resolution input

LeHenschel committed:
Removed unnecessary print statement

LeHenschel committed:
Added FastSurferVINN model as default

LeHenschel committed:
Example .yaml file for definition of training run

LeHenschel committed:
Merge branch 'feature/vinn' of https://github.com/Deep-MI/FastSurfer into feature/vinn

LeHenschel committed:
Fixed wrong model name

LeHenschel committed:
Changed default config assignment to example yamll-file

LeHenschel committed:
Updated conform to newest version

LeHenschel committed:
Added saving and loading routines including conform.py calls

LeHenschel committed:
Modularization of labels. Can pass look-up table instead of hard-coded labels

af-a committed:
Fix get_cfg_defaults import in load_config.py

af-a committed:
Add missing FSCNN deps in conda yml files

af-a committed:
Update reqs.in and compiled reqs.txt with missing FSCNN deps

af-a committed:
[data_utils] Modify map_label2aparc_aseg to process torch Tensors

Fixes runtime error in run_prediction due to torch functions being
applied to a np.ndarray.
Also corrected labels variable type.

af-a committed:
[run_prediction] Fix call: save_image_as_nifti --> save_image

af-a committed:
Add fn for downloading ckpts in checkpoint.py

af-a committed:
Add download_check_point.py util script

af-a committed:
Update hdf5 args and examples in FSCNN README

af-a committed:
Update training description and examples in FSCNN README

af-a committed:
[run_prediction] Add conformed image saving

af-a committed:
[run_prediction] Modify conf. image saving routine

Includes modifying the input to set_subject (now the full subject path
instead of just a name) and setting the current conf_name within the function.

af-a committed:
[run_prediction] Add original input image saving

af-a committed:
[run_prediction] Add in_dir arg; alternative to csv_file

af-a committed:
[run_prediction] Add missing checks for input and output options

af-a committed:
[run_prediction] Save preds in out_dir instead of input subject dirs

af-a committed:
Modifications for single_img case

* Interpret orig_name, gt_name, and pred_name as absolute paths (as in previous version).
* Adapting conf_name and input_img_names according to single_img

af-a committed:
[run_prediction] Get image info from orig instead of gt data

af-a committed:
[run_prediction] Add image conforming (updated fpr hires)

af-a committed:
[run_prediction] Add option of running view agg on cpu or gpu

LeHenschel committed:
Comparison labels from list and labels present in segmentation added

af-a committed:
[run_prediction] Remove extra function comments

af-a committed:
[run_prediction] Change output dir in single_img case to pred_name dir

LeHenschel committed:
Changed addition to assignment for sagittal view only run

af-a committed:
[conform] Fix argument parsing errors

af-a committed:
Modify save_img function to save segs with type int16

Solves a later problem with recon-surf (Segmentation fault in the CortexLabel stage).

dkuegler committed:
[GPU memory optimization/interpolation_layer] Updated the Zoom interpolation_layers to use no torch.cat if only using one scale-factor setting (specifically for inference) and add support for changing the output shape

af-a committed:
[run_prediction] Activate denormal flushing in FastSurferCNN

af-a committed:
Update run_fastsurfer.sh to use updated inference script

af-a committed:
Add no_cuda option in main and inference scripts

dkuegler committed:
[GPU memory optimization/model] Change model, eval script to use the optimal target shape (no extra work), automatically infer where to do inference by device of the prediction volume, simplify run_prediction.py, add torch.no_grad() to all inference-only functions (the latter is the "real solution" to the memory usage problems)

dkuegler committed:
[GPU memory optimization/model] Change model, eval script to use the optimal target shape (no extra work), automatically infer where to do inference by device of the prediction volume, simplify run_prediction.py, add torch.no_grad() to all inference-only functions (the latter is the "real solution" to the memory usage problems)

dkuegler committed:
[data_utils] fix and optimize split_cortex_labels

dkuegler committed:
[eval] make script flexible to differently sized images (padded or not)

dkuegler committed:
[data_utils] cleanup the map_label2apac_aseg function

dkuegler commented:
@AhmedFaisal95 / @LeHenschel 

In the current feature/vinn https://github.com/Deep-MI/FastSurfer/pull/168/commits/03c0af0c7d477d1097b5f7c58c418ed2d6ee8946 If I just pass the sagittal cfg and checkpoint, it crashes

> Traceback (most recent call last):
>   File "FastSurferCNN/run_prediction.py", line 434, in <module>
>     pred_data = eval.get_prediction() #
>   File "FastSurferCNN/run_prediction.py", line 237, in get_prediction
>     pred_prob = self.run_model(pred_prob)
>   File "FastSurferCNN/run_prediction.py", line 165, in run_model
>     pred_prob = self.model.run(self.orig_filename, self.orig_data, self.orig.header.get_zooms(),
>   File "/home/kueglerd/projects/FastSurfer/FastSurferCNN/eval.py", line 172, in run
>     pred_prob = self.eval(test_data_loader, pred_prob, out_scale=out_res)
>   File "/home/kueglerd/projects/FastSurfer/venv/lib/python3.8/site-packages/torch/autograd/grad_mode.py", line 28, in decorate_context
>     return func(*args, **kwargs)
>   File "/home/kueglerd/projects/FastSurfer/FastSurferCNN/eval.py", line 154, in eval
>     pred_prob[start_index:start_index + pred.shape[0], :, :, :] += torch.mul(pred, 0.2)
> RuntimeError: The size of tensor a (51) must match the size of tensor b (79) at non-singleton dimension 3

This is due to the output `pred` being initialized with the model's class number (51) but this tries to save it into the view_aggration tensor `pred_prob` which is (79) classes.

af-a commented:
> @AhmedFaisal95 / @LeHenschel
> 
> In the current feature/vinn [03c0af0](https://github.com/Deep-MI/FastSurfer/commit/03c0af0c7d477d1097b5f7c58c418ed2d6ee8946) If I just pass the sagittal cfg and checkpoint, it crashes
> 
> > Traceback (most recent call last):
> > File "FastSurferCNN/run_prediction.py", line 434, in 
> > pred_data = eval.get_prediction() #
> > File "FastSurferCNN/run_prediction.py", line 237, in get_prediction
> > pred_prob = self.run_model(pred_prob)
> > File "FastSurferCNN/run_prediction.py", line 165, in run_model
> > pred_prob = self.model.run(self.orig_filename, self.orig_data, self.orig.header.get_zooms(),
> > File "/home/kueglerd/projects/FastSurfer/FastSurferCNN/eval.py", line 172, in run
> > pred_prob = self.eval(test_data_loader, pred_prob, out_scale=out_res)
> > File "/home/kueglerd/projects/FastSurfer/venv/lib/python3.8/site-packages/torch/autograd/grad_mode.py", line 28, in decorate_context
> > return func(*args, **kwargs)
> > File "/home/kueglerd/projects/FastSurfer/FastSurferCNN/eval.py", line 154, in eval
> > pred_prob[start_index:start_index + pred.shape[0], :, :, :] += torch.mul(pred, 0.2)
> > RuntimeError: The size of tensor a (51) must match the size of tensor b (79) at non-singleton dimension 3
> 
> This is due to the output `pred` being initialized with the model's class number (51) but this tries to save it into the view_aggration tensor `pred_prob` which is (79) classes.

I encountered this in my initial tests (since https://github.com/Deep-MI/FastSurfer/pull/168/commits/60e7c6d71f2cb94484088f10bcbbd35585abc27e) and discussed it. From what I understood,
FastSurfer is not usually run in the single-plane case and so it is not explicitly covered, but @LeHenschel can correct me if I am wrong.

dkuegler committed:
[run_prediction] VERIFIED ON BERT; simplify get_prediction memory allocation rules, add get_num_classes, fix single-plane prediction

dkuegler committed:
[run_prediction] VERIFIED ON BERT; simplify get_prediction memory allocation rules, add get_num_classes, fix single-plane prediction

dkuegler commented:
> I encountered this in my initial tests (since [60e7c6d](https://github.com/Deep-MI/FastSurfer/commit/60e7c6d71f2cb94484088f10bcbbd35585abc27e)) and discussed it. From what I understood, FastSurfer is not usually run in the single-plane case and so it is not explicitly covered, but @LeHenschel can correct me if I am wrong.

that makes sense, but I am trying to hunt down why my predictions with the memory optimizations are different and I basically have narrowed it down to the view aggregration so I need single-plane prediction to detect the bugs.

af-a commented:
> > I encountered this in my initial tests (since [60e7c6d](https://github.com/Deep-MI/FastSurfer/commit/60e7c6d71f2cb94484088f10bcbbd35585abc27e)) and discussed it. From what I understood, FastSurfer is not usually run in the single-plane case and so it is not explicitly covered, but @LeHenschel can correct me if I am wrong.
> 
> that makes sense, but I am trying to hunt down why my predictions with the memory optimizations are different and I basically have narrowed it down to the view aggregration so I need single-plane prediction to detect the bugs.

I see. I think either coronal or axial on their own would work.

dkuegler commented:
https://github.com/Deep-MI/FastSurfer/pull/177 a pull request to pull into a pull request :-)

dkuegler committed:
Merge pull request #177 from dkuegler/feature/vinn-fix-single-plane

[run_prediction] VERIFIED ON BERT; simplify get_prediction memory all…

dkuegler committed:
Merge pull request #175 from dkuegler/feature/fix-warning

[data_utils] fix and optimize split_cortex_labels

dkuegler committed:
[run_prediction] move the set_model commands to be more "reliable"

dkuegler commented on FastSurferCNN/run_prediction.py:
```diff
@@ -0,0 +1,419 @@
+            # (this currently takes only the total memory into account, not the occupied on)
+            total_gpu_memory = sum([torch.cuda.get_device_properties(i).__getattribute__("total_memory") for i in
+                                    range(torch.cuda.device_count())])
+            self.small_gpu = total_gpu_memory < 8000000000
```
this number needs to be adjusted for VINN, it now depends on the size of the image:
formula for memory required for view aggregation:
`width (up to 320) * height (up to 320) * depth (up to 320) * classes (79) * sizeof(float) (4)` which is about 9.64 GB

dkuegler committed:
[run_prediction/eval/interpolation_layer] fix the memory issue, optimize view aggregation and interpolation layer, introduce half precision for view aggregation

dkuegler committed:
[interpolation_layer] Fix/Error message, if passing bad arguments (scale_factors) to forward

LeHenschel committed:
Hires-flag for hires processing in recon-surf (#181)

* Hires-flag for hires processing in recon-surf

* Hires option for VINN. By default, images are conformed to 1.0 mm before running the DL network

LeHenschel committed:
Merge pull request #176 from dkuegler/feature/vinn

Fix the memory issues of the VINN branch

LeHenschel committed:
Fixed class assignment issue - call class function get_num_classes, not get method of model

LeHenschel committed:
Merge branch 'dev' into feature/vinn

m-reuter committed:
Merge branch 'dev' into feature/vinn

LeHenschel committed:
Optional look-up table to infer re-mappings from

LeHenschel committed:
Merge branch 'feature/vinn' of https://github.com/Deep-MI/FastSurfer into feature/vinn

LeHenschel committed:
Update config files for running and training FastSurferVINN

LeHenschel committed:
Merge branch 'feature/vinn' of https://github.com/Deep-MI/FastSurfer into feature/vinn

LeHenschel committed:
Set default config files and make checkpoints and cfgs non-optional

LeHenschel committed:
Added default ckpts and download link

LeHenschel committed:
corrected default settings (required status and path) for cfg and ckpts

LeHenschel committed:
Fixed format issues and removed depreceated element

LeHenschel committed:
Fixed ckptdir split into dir and basename

m-reuter committed:
added denormal flushing

LeHenschel committed:
Changed no_cuda to device flag

dkuegler commented on FastSurferCNN/run_prediction.py:
```diff
@@ -379,7 +379,10 @@ def check_and_download_ckpts(ckptpath):
 
-    parser.add_argument('--no_cuda', action='store_true', default=False, help="Disables GPU usage")
+    parser.add_argument('--no_cuda', help="Deprecated: Use --device flag instead!")
+    parser.add_argument('--device', default="cuda",
```
There should probably an "auto" option for device, which should be the default.

This is so the default configuration works independent of the hardware of the PC.

I think we also had some automatic identification of Hardware in fastsurfer before.

dkuegler replied:
I would even go as far as raising an exception, if cuda is passed as the device, but torch does not run on the GPU (e.g. the device is invalid)

dkuegler replied:
Similarly, batch_size should probably also default to some automatic behavior if no number is given to ensure the default configuration most likely works.

I am going to create a dedicated issue for this, since this is not really required for v2.0 and can be done in a later update.

LeHenschel replied:
Currently, we just switch to cpu if gpu is set, but not available and return a warning. Assertion works as well. For auto, this only makes sense in my opinion once we include the other options (MAC M3 and what not). @m-reuter wanted to create a check list for that.

dkuegler replied:
My idea is that if you specifically specify cuda, you should get an error if it can't do it on the GPU.

This is a clear expected behavior that also lets you debug well. 

See also https://peps.python.org/pep-0020/

LeHenschel replied:
Yeah, I am not arguing against it. I would just change it in one go with the auto option.

m-reuter replied:
Reply I updated it to "auto" so it will be easier in the future to add more options.

LeHenschel committed:
Add actual assignment of device to cpu, when cuda not available

LeHenschel committed:
Added device flag and set no-cuda to deprecated

m-reuter committed:
updated --device flag

m-reuter committed:
update checkpoint downloading and command line install script

m-reuter committed:
update Dockerfile to download checkpoints and new entry point run_prediction

m-reuter committed:
moved to FastSurferCNN/download_checkpoints.py

m-reuter commented:
I just merged and updated the default checkpoints download function and command line script. And I updated the dockerfiles to download the checkpoints and to point to the correct entry script for the Segmentation only ones: run_prediction. We need to consider always using run_fastsurfer as entry point to avoid these kinds of breaking changes when network scripts get changed.

m-reuter committed:
Merge branch 'dev' into feature/vinn

dkuegler commented:
should we also prune the FastSurferCNN checkpoints out of the git tree with the 2.0 release then?

m-reuter committed:
added info on downloading checkpoints

m-reuter commented:
> should we also prune the FastSurferCNN checkpoints out of the git tree with the 2.0 release then?

Probably not, as this will break things if someone checks out an older release tag. Maybe in the distant future. But they can be deleted and downloaded from now on.

dkuegler committed:
optimize interpolation_layer to be skipped if no interpolation is necessary (#188)

m-reuter committed:
update --device flag

m-reuter committed:
add mps

m-reuter committed:
fix paths in checkpoint files and skip ssl check, as fz-juelich certificate seems to have expired

dkuegler committed:
cleanup the logging in run_prediction.py (#189)

- replace stub get_logger to getLogger
- add __init.py in utils
- rename logging_script.py to logging.py (which is now an extension to the builtin logging, so it is less confusing)
- remove duplicate logging
- move the initialization of the logger to the beginning (instead of in __init__ of RunModelOnData)
- added a tqdm progress bar and cleaned log messages for batches in Inference.eval
- updated get_affine() to .affine in one occurrence in run_prediction.py

Co-authored-by: LeHenschel <leonie.henschel@dzne.de>
Co-authored-by: Martin Reuter <m-reuter@users.noreply.github.com>

m-reuter committed:
workaround for broken chpt mps loading and avoid loading coronal twice

LeHenschel committed:
Fix: checkpoints directory, simplified: run_prediction call

LeHenschel committed:
Removed unused code snippets, restructured to simplified version, removed --save_img and --single_img routines (automatically infered)

m-reuter committed:
fix: really map labels to mps

m-reuter committed:
bump h5py and numpy to match mac

m-reuter committed:
updated mac requirements to VINN and mps (nightly torch)

af-a committed:
Revert "Download correct conda architecture"

This reverts commit 300a1c552c9c37391a6bff34837e2b005891f018.

af-a committed:
Include --vinn flag in download_checkpoints.py calls (dockerfiles)

Otherwise, this will always fail due to the check:
"Specify either files to download or --vinn"

LeHenschel commented on FastSurferCNN/eval.py:
```diff
@@ -109,8 +110,18 @@ def set_model(self, cfg=None):
+        else:
+            model_state = torch.load(ckpt, map_location=self.device)
+            self.model.load_state_dict(model_state['model_state'])
+        #print(model_state['model_state']['inp_block.bn0.weight'])
```
Remove the print statements?

m-reuter replied:
Ok, can you do this.

LeHenschel reviewed (commented):
Shouldn't we rather change the download script in this case? VINN should be the default and defining all and vinn is a bit counter intuitive

LeHenschel commented on FastSurferCNN/download_checkpoints.py:
```diff
@@ -0,0 +1,50 @@
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description='Check and Download Network Checkpoints')
+    parser.add_argument("--all", default=False, action="store_true",
```
I would rather make this a flag with which you can provide different arguments (all, vinn, cnn) and use all as the default

dkuegler replied:
Currently, vinn should be the only and automatic option. If people want to use faalstsurfercnn, they should use fastsurfer 1.X imo
If we integrate other modules for other task, it should be a list of modules (e.g. aseg, aparc)

m-reuter replied:
I want an all flag to download everything that is available. This will be used in docker. 
Integration of fscnn well be decided based on vinn performance in tests.

LeHenschel replied:
My main issue here is that there should not be multiple flags (--all --vinn --cnn). Especially because at the  moment you need to specify both (--all and --vinn). "all" should already include "vinn" at the very least. Otherwise it is very counterintuitive.

...and faalstsurfercnn is pretty hilarious (=failiest in dutch).

m-reuter replied:
all does include vinn, see the fixes below. I agree that we need to think about having a --flag for each module later or having a --module string or list. I think there is really no difference here. What is more important is to think about good names for the modules. vinn is not great as many module will use the vinn architecture. Probably should be fsvinn for FastsurferVINN . Names cannot be too long.

LeHenschel replied:
Ah, ok. Did not see the new commit. 

Could also be a combination of network and structure (similar to what david suggested --arch vinn and --module aseg, aparc, olfbulb...). Downside is that probably only certain combinations are possible. I get your point and agree that we need a better description, but FastSurfer (therefore fsvinn and FastSurferVINN) are in my opinion not exclusive to the aparc segmentation (just like you also have hippo/thalamus submodules in FreeSurfer).

m-reuter replied:
another downside of the arch and module is that people would not usually know what arch was used for a task. They just want a specific module and don't care what arch that uses. Which is why we should try to only have one candidate for a specific task. Maybe it should be aparcDKT , cerebellum, olfactorybulb, hypothalamus, choroidplexus, piriformcortex etc.

LeHenschel replied:
Sounds good. Could also just have a default and for people who want to try it out can change it (depending if there actually are different architectures for one submodule).

dkuegler replied:
Naming of modules should be consistent with output segmentation file names (so aseg+aparcDKTatlas)

m-reuter committed:
fix all flag

m-reuter committed:
download checkpoints all includes vinn now

m-reuter committed:
Add experimental docker for AMD GPUs

m-reuter committed:
Merge branch 'dev' into feature/vinn

af-a committed:
[eval] Remove network config printing in logs

af-a committed:
[logging] Modify the setup_logging function

The function now accepts a single argument: log_file_path.
If an empty string is provided, logging messages are only printed on stdout
and not saved to a file.

af-a committed:
[run_prediction] Add optional arg log_name; adapt setup_logging call

af-a committed:
[run_fastsurfer] Add log_name arg to run_prediction call; remove output pipe

m-reuter committed:
update mps text for apple

m-reuter committed:
bump to FS7.3.2 and Ubuntu 20

m-reuter committed:
bump freesurfer to 732

m-reuter committed:
added libquadmath0 needed by talairach avi in fs732

m-reuter committed:
add fs_temp_file and fs-check-version, remove dev rca_config download

m-reuter committed:
sort lists alphabetically

af-a committed:
[run_fastsurfer] Correct reference in log message: eval.py --> run_prediction.py

af-a committed:
[checkpoint] Add cert verification and a fallback option (no verification)

In addition, the script currently prints out the response code if
an HTTPError is raised.

af-a committed:
[checkpoint] Add missing raise_for_status() following second url request

dkuegler commented on FastSurferCNN/utils/checkpoint.py:
```diff
@@ -142,6 +142,7 @@ def download_checkpoint(download_url, checkpoint_name, checkpoint_path):
     except requests.exceptions.HTTPError as e:
         print('[checkpoint] Response code: {}'.format(e.response.status_code))
```
Should this not be logged so it is also in the log files?

af-a replied:
Yes. We can add a LOGGER instance in `checkpoint.py` to log this message as well.

af-a committed:
Merge pull request #194 from AhmedFaisal95/fix/inference-logging

Fix: Inference script logging

af-a committed:
[checkpoint] Add logging of info message

LeHenschel committed:
Rename eval to inference.py

LeHenschel committed:
Remove print statements

LeHenschel committed:
Correct import from inference after renaming

LeHenschel committed:
Changed default checkpoints name to download into FastSurferCNN/checkpoints directory

LeHenschel committed:
Arguments matched to run_fastsurfer.sh naming scheme

LeHenschel committed:
Set dest for old orig_dir var

LeHenschel committed:
Naming of run_prediction arguments to match new scheme

LeHenschel committed:
Naming of conformed image name argument to match new scheme

af-a committed:
[recon-surf] Add check and error for very long sid/subject paths

af-a committed:
[recon-surf] Raise error if running on existing subject directories

m-reuter committed:
added script for checking segmentation

LeHenschel committed:
Fix: Missing subjectID when strip flag is not set

LeHenschel committed:
Updated ckpts to new naming scheme and url

m-reuter committed:
fix binpath, time on mac and check segmentation

m-reuter committed:
added details on AMD and Mac

m-reuter committed:
fix typo

m-reuter committed:
update and

m-reuter committed:
Adjust failure threshold up to .77

af-a committed:
[run_fastsurfer] Fix config_cor, config_ax args

m-reuter committed:
fix docker pull and run names

af-a committed:
Move quick_qc.py: recon_surf --> FastSurferCNN

Allows importing check_volume within run_prediction.py in FastSurferCNN/.

af-a committed:
[quick_qc] Modify check_volume to return bool, not exit within function

af-a committed:
[run_prediction] Add volume-based QC check

If QC check fails for a single subject: exit with error.
If it fails for a batch of subjects: report the number of failures.

af-a committed:
[run_fastsurfer] Add error message indicating seg qc failure

dkuegler committed:
optimize N4 bias field correction:
- replace repeated percentiles (sorting) with single call
- remove some of the repeated conversion between datatypes (unit/float32)

dkuegler committed:
replace skimage erosion/dilation with scipy.ndimage

af-a committed:
[run_prediction] Add optional output: log file listing QC-failed subjects

dkuegler committed:
Interface changes in preparation for modules.
Cleanup of the import statements.
Change of the "base path" of FastSurferCNN (now the include statements start with FastSurferCNN), i.e. the python search path should include the FastSurfer project folder and not the FastSurferCNN subfolder.
Move parameters common across different modules to FastSurferCNN/utils/parser_defaults.py
Simplify/unify some of the run_prediction interface/internal structure
Update type annotation (2x) in conform.py
Move the auto-detection of the device out of the main inference script.
clean up of the call chain and arguments in FastSurferCNN/inference.py

af-a committed:
[run_prediction] Substitute QC-failed subject list for simple count

af-a committed:
[run_prediction] Remove extra try-except block

af-a committed:
[run_prediction] Remove extra comment

af-a committed:
[run_fastsurfer] Rectify default weight file names

af-a committed:
[recon-surf] Adapt cmd to modified quick_qc location

m-reuter committed:
support negative values in image (UKbio) (#210)

* support negative values in image (UKbio)
* also setting values to zero if they are perfectly zero in input

m-reuter committed:
merge dev into VINN branch

LeHenschel committed:
Added arguments for aseg_name and brainmask_name

Removed aseg and brainmask prediction from recon-surf (now in run_prediction)

Added brainmask prediction and aseg generation. Changed savind routine in eval to allow uint8 saving.

Copy files from recon-surf, export FASTSURFER_HOME, fix import error with new module structure

Export python path for checkpoint download

Printing subject name for debugging

Subject ID as separate flag for single image run: add argument, update set_subject functionality and pass sid inside run_prediction.py call

Subject ID as separate flag for single image run: pass correct variable in run_fastsurfer

Removed debugging logger statement

make checkpoints word readable

af-a committed:
Fix spacing in some conditional statements (run_fastsurfer.sh and recon-surf.sh)

af-a committed:
[recon_surf] Fix mris_place_surface arg: aparc_aseg_segfile --> seg

LeHenschel committed:
world readable

af-a committed:
[recon_surf] Fix mri_segstats args

* aparc_aseg_segfile --> seg
* brain-vol-from-aparc_aseg_segfile --> brain-vol-from-seg

af-a committed:
Move .dockerignore to root directory

LeHenschel committed:
Changed MGH to SpatialImage in conform.py, moved reduce_to_aseg.py into FastSurferCNN module and adapted run_prediction import statement

LeHenschel committed:
Delete reduce_to_aseg.py in recon-surf
